engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -255,99 +332,149 @@ private Void handleCreateVolumeFromSnapshotBothOnStorageSystem(SnapshotInfo snap
+    // create a volume from the snapshot so that we can copy the VHD file that should be inside of the snapshot to secondary storage.
+    //
+    // The resultant volume must be writable because we need to resign the SR and the VDI that should be inside of it before we copy
+    // the VHD file to secondary storage.
+    //
+    // If the storage system is using writable snapshots, then nothing need be done by that storage system here because we can just
+    // resign the SR and the VDI that should be inside of the snapshot before copying the VHD file to secondary storage.
+    private void createVolumeFromSnapshot(HostVO hostVO, SnapshotInfo snapshotInfo, boolean keepGrantedAccess) {
+        SnapshotDetailsVO snapshotDetails = handleSnapshotDetails(snapshotInfo.getId(), "tempVolume", "create");
 
-            copyCmdAnswer = (CopyCmdAnswer)_agentMgr.send(hostVO.getId(), copyCommand);
-        }
-        catch (Exception ex) {
-            throw new CloudRuntimeException(ex.getMessage());
+        try {
+            snapshotInfo.getDataStore().getDriver().createAsync(snapshotInfo.getDataStore(), snapshotInfo, null);
         }
         finally {
-            try {
-                _volumeService.revokeAccess(snapshotInfo, hostVO, snapshotInfo.getDataStore());
-            }
-            catch (Exception ex) {
-                s_logger.debug(ex.getMessage(), ex);
-            }
-
-            try {
-                _volumeService.revokeAccess(volumeInfo, hostVO, volumeInfo.getDataStore());
-            }
-            catch (Exception ex) {
-                s_logger.debug(ex.getMessage(), ex);
-            }
+            _snapshotDetailsDao.remove(snapshotDetails.getId());
         }
 
-        String errMsg = null;
+        CopyCmdAnswer copyCmdAnswer = performResignature(snapshotInfo, hostVO, keepGrantedAccess);
 
         if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
             if (copyCmdAnswer != null && copyCmdAnswer.getDetails() != null && !copyCmdAnswer.getDetails().isEmpty()) {
-                errMsg = copyCmdAnswer.getDetails();
+                throw new CloudRuntimeException(copyCmdAnswer.getDetails());
             }
             else {
-                errMsg = "Unable to perform host-side operation";
+                throw new CloudRuntimeException("Unable to perform host-side operation");
             }
         }
+    }
 
-        CopyCommandResult result = new CopyCommandResult(null, copyCmdAnswer);
+    // If the underlying storage system needed to create a volume from a snapshot for createVolumeFromSnapshot(HostVO, SnapshotInfo), then
+    // this is its opportunity to delete that temporary volume and restore properties in snapshot_details to the way they were before the
+    // invocation of createVolumeFromSnapshot(HostVO, SnapshotInfo).
+    private void deleteVolumeFromSnapshot(SnapshotInfo snapshotInfo) {
```

Comment:
@mike-tutkowski I think that use Javadoc instead of comment here might be interesting.